feat: add visual editor

selvalt7 opened the pull request:
#773 
![editor](https://github.com/user-attachments/assets/ad326b8c-99c5-421c-9f45-b2e164ea0150)
![thresholds](https://github.com/user-attachments/assets/6d459e7b-5481-4adb-b465-9ce1cce775e2)
![entity](https://github.com/user-attachments/assets/b835e366-3a85-4c2a-84a7-ea11497ba86e)
![appearance](https://github.com/user-attachments/assets/55b9d4e2-3177-4aea-abca-7d5e6d869589)

selvalt7 committed:
ok

selvalt7 committed:
Add editor

selvalt7 committed:
Fix

selvalt7 committed:
Improved visual editor

selvalt7 committed:
Update naming

selvalt7 committed:
Added color selector
Added translations

selvalt7 committed:
Add visual editor

selvalt7 committed:
Fix

selvalt7 committed:
Fix: color conversion functions

selvalt7 committed:
Added support for color array
Fixed not able to add entity in a new card
Added option whether to use color thresholds

selvalt7 committed:
Added color thresholds editor

selvalt7 committed:
Set default show values
Added headers to sub pages
Css adjustments
Added more settings
Entity form now uses grid for selectors positioning

selvalt7 committed:
Removed use color thresholds option
Added clear button to color selector
Color thresholds are in expansion panels now
Added support for entity as value

selvalt7 committed:
Added state map editor
Added subpage header and subpage link elements
Organized files
Added few default values
Added bounds and value factor options
Color selector small css tweak

selvalt7 committed:
Made subpage link focusable
Empty color selector is now indicated by a checkerboard

selvalt7 committed:
Removed color conversion functions
Removed support for color rgb array

selvalt7 committed:
Merge branch 'dev' of https://github.com/selvalt7/mini-graph-card into dev

selvalt7 committed:
Merge branch 'dev' into editor

selvalt7 committed:
ok

akloeckner commented:
Wow! Very impressive. I will need some time to dive into this. And I will most likely need a few iterations until I understand everything.

If you appreciate some early feedback from my browsing around for the first time:
* I found this glitch in the "Display" submenu, which does not happen in the "Alignment" submenu: 
  <img width="301" alt="image" src="https://github.com/user-attachments/assets/0d21d0ee-b282-4ccd-bb0d-25da6bff637e">

* We are having some issues with the `package-lock.json` file. It loses `resolved` and `integrity` fields, whenever a new package is installed. You also ran into this issue. The only fix I could find so far, was to manually select the changes to commit and revert all the others.

* The entities show two buttons to delete them:
  <img width="310" alt="image" src="https://github.com/user-attachments/assets/0881e92b-3b77-4e21-9cfa-e9fbd068bfed">
  Maybe we can manage to show and use the default one only?

selvalt7 committed:
fix: missing "resolved" and "integrity" field in "package-lock.json" file

selvalt7 committed:
fix: visual glitch in "select" selector

selvalt7 committed:
feat: entity can be removed with button inside the selector

Removed the other button for removing entity

selvalt7 committed:
fix: unset show properties were treated as undefined

ildar170975 commented:
Just a small remark:
For some STOCK cards a UI does not provide controls for changing some options - i.e. these options may only be changed in yaml. There are some related issues in HA frontend Github - and some of these issues are closed be Devs as "well, it turned to be normal to force users to use yaml to define some options", some issues may be just ignored.
Means - may be we should not try to support ALL available options. Besides, one day some options may become per-entity - and then changes in UI will be needed.
I am really satisfied that @selvalt7 started developing UI - but I do not insist on a support of ALL available options.

ildar170975 commented:
Another remark is about selecting entities.
Since I have not tested is PR - have to assume & guess about a possible use of "common HA ways" in this PR.
The current "common way" of selecting entity - and then displaying a selected entity - is rather confusing & unfriendly:
1. After being selected - an entity is shown by `friendly_name` which can be very confusing since names are not unique. Check https://github.com/home-assistant/frontend/issues/11870 for Automations UI; similarly it is about [editing Entities card](https://github.com/home-assistant/frontend/issues/11938#issuecomment-2304549186) (& all similar places in HA).
2. When a drop-down UI control is reopened - the selected entity is not "selected" (https://github.com/home-assistant/frontend/issues/17395) - which makes hard to guess which entity is selected (unless you open a yaml).

Means - using same "common ways" just "contaminate" more cards in HA ))).
Not a big deal, using yaml anyway - just a warning.

selvalt7 commented:
Visual editor does not provide all the options only the ones I felt are necessary, if user wants to change card size or change update interval to something other than every state change, it can only be done with yaml.

On the subject of selected entity which is represented with `friendly_name`. I came up with an idea to display `entity_id` under the selector like this 
![entity selector](https://github.com/user-attachments/assets/4a31c1a1-143b-448b-bd9a-275cf24054c3)

ildar170975 commented:
> I came up with an idea to display entity_id under the selector like this

Good, very useful & makes a picture clear.

And what about this [common issue](https://github.com/home-assistant/frontend/issues/17395)?
I was afraid that you will face same glitch if decide to use a common component.

selvalt7 commented:
Yeah I have that issue, the only solution I think is to make custom selector or wait until they fix it in entity selector.

selvalt7 committed:
feat: display entity_id under entity selector

This should help identify selected entity as the friendly_name is not unique and the default entity selector does not correctly highlight selected entity

ildar170975 commented:
> wait until they fix it in entity selector.

I would choose this way for time-saving.

selvalt7 committed:
Merge branch 'dev' into editor

akloeckner commented on src/editor/components/colorSelector.js:
```diff
@@ -0,0 +1,127 @@
+  hex_color: {},
+};
+
+export class CustomColorSelector extends LitElement {
```
Is there a reason not to use the color picker provided by hass itself?

If it is the "clearable" feature, it could be handled in the usual hass way by having an enable/disable toggle in front of the color picker.

If it is the hex format, could it be converted "output-only" in the main `valueChanged` function? I am not familiar enough with the way these things work...

selvalt7 replied:
Mainly it was the hex format. I've tried to do the convertion on output but this resulted in warnings in console about supplied color value to hass color picker but the selecter was rendering just fine.

akloeckner replied:
Ah, I see... I'd guess, we'd have to convert back and forth to not loose colours on each startup of the editor... At least in `setConfig`, `configChanged` *and* `entitiesChanged`... Then use `[r,g,b]` internally in the editor and hex outside... 

Another idea: How about we allow `[r,g,b]` as a new colour input format for the mini-graph-card itself? Better suited for a follow-up PR then, maybe... We'd still loose non-hex colours on first startup of the editor, I assume... 

I shall test, if losing colours is actually an issue...

akloeckner replied:
I did a test, and yes, loosing colours is an issue. Try this: 
* Edit card
* Switch to YAML
* Set colour to "yellow" on an entity
* Switch to UI mode and edit entity
* Observe the colour to be black now.

selvalt7 replied:
I've addressed this issue in the latest commit

akloeckner commented on src/editor/editor.js:
```diff
@@ -0,0 +1,561 @@
+    if (!this._config || !this.hass) {
+      return;
+    }
+    // this._config = { config, ...this._entities };
```
Is this a left-over comment?

akloeckner commented on src/main.js:
```diff
@@ -127,6 +133,14 @@ class MiniGraphCard extends LitElement {
+    return {
+      entities: [],
+      color_thresholds: [],
+      show: DEFAULT_SHOW,
```
I do not understand, when this is used. If it automatically adds the default show options to each new card, we should not do this, I think. If it just can be used by the user to get a sensible template to fill in, I think it is a good idea to return all these otions here...

selvalt7 replied:
This was needed because things like adding new entity on a new card was now working properly but I sorted it out internally so it is not needed anymore and will be deleted.

akloeckner reviewed (commented):
I finally got around to dig into some of the details of your (still impressive!) PR. Thanks for hanging on. I might have more feedback later, but I thought I'd better share it sooner than later.

To sum up my thoughts up front in a few points: I found one or two minor glitches. And I think we can improve some of the code by re-using common patterns and facilities provided by hass itself, which may also help to increase similarity to other cards' config editors.

Please have a look and let me know what you think. I might also not have grasped the full picture yet.

selvalt7 commented on src/editor/editor.js:
```diff
@@ -0,0 +1,561 @@
+    this.subElementEditorConfig = { type: 'statemap', index: 0 };
+  }
+
+  renderSubElement() {
```
Thanks for the suggestion, maybe instead of subpages, `color_thresholds` and `state_map` whould be in their respective expansion panel under the appearance

akloeckner replied:
I did feel a bit unfamiliar with the concept of subpages, I must admit. However, I don't think, the entities' option can easily be folded into some expansion panel. So, we'd have subpages anyways, I guess. Then, why not use them consequently? Unless we find a way to also cascade the entities' options in another way...

My point here was in a slightly different direction: as both state map and colour thresholds create a list of schema-defined entries, this functionality could maybe be generalized and reused twice. Instead of the code being repeated (with some adaptations) twice.

selvalt7 replied:
Yeah, the entities subpage is here to stay, but I thought about changing `color_thresholds` and `state_map` to use the custom `ha-form` element so it can be used in the main schema like this:
```js
 {
"the rest of the schema"
 },
 {
   name: 'color_thresholds',
   type: 'expandable',
   iconPath: mdiFormatColorFill,
   schema: [
     {
       type: 'mgc-list',
       schema: [
         {
           name: '',
           type: 'grid',
           column_min_width: '100px',
           schema: [
             {
               name: 'value',
               selector: { number: { step: 0.1 } },
             },
             {
               name: 'color',
               selector: { hex_color: {} },
             },
           ],
         },
       ],
     },
   ],
 },
```
That results in this:
![editor](https://github.com/user-attachments/assets/2f2bd882-3aba-4b4e-afb2-69be921cd941)

akloeckner replied:
Thanks for clarifying. That was exactly my suggestion as well. :-)

selvalt7 commented on src/localize/localize.js:
```diff
@@ -0,0 +1,27 @@
+  }
+}
+
+export default function localize(key, hass) {
```
I think I have found a way to add custom translations for hass.localize.
Basically you have to add new property to hass.resources but I don't know if it's right way to do it.

akloeckner replied:
Nice catch! It seems they even have a private function to do this... I haven't found a public one yet...

https://github.com/home-assistant/frontend/blob/fd9e2b647debc580b56354c2344837d594c43702/src/state/translations-mixin.ts#L411

I'll try and search some more, when the rest of the world is asleep here... ;-)

selvalt7 committed:
Added `ha-form-mgc-list` for use in `ha-form` schema
Color thresholds and state map now use the new `ha-form-mgc-list`
Removed old unused components
Moved schemas to `editorConst.js`
Translations are now handled through `hass.localize`
Color selector should now display proper color when supplied with color name
Fixed inability to clear few options
Removed leftover comments

tr05t committed:
add German translation

tr05t committed:
add German translation

tr05t committed:
Merge branch 'editor' of github.com:tr05t/mini-graph-card into editor

tr05t committed:
add German translation

tr05t committed:
Merge branch 'editor' of github.com:tr05t/mini-graph-card into editor

selvalt7 committed:
add German translation

selvalt7 committed:
Added value localization
Added missing translations

justinkruit commented:
Any reason this hasn't been merged?

ildar170975 commented:
> Any reason this hasn't been merged?

Every PR must be reviewed first before a merge. And there are other PRs which should be reviewed  first as more prioritized. Adding UI is definitely not as important as fixing bugs.

Liquidmasl commented:
> well, it turned to be normal to force users to use yaml to define some options", some issues may be just ignored.
> Means - may be we should not try to support ALL available options. Besides, one day some options may become per-entity - and then changes in UI will be needed.

the HA team goes towards UI stuff more though, and wants to phase out yaml configs as far as possible. But sure some edge features may stay with yamls. that being said, I dont think it should ever be considered not bringing something into the ui if it can indeed be brought into the ui. 
expecially cause the visual editor can be SO MUCH MORE powerful then any cards (that i know of) actually make use off haha

ildar170975 commented:
> though, and wants to phase out yaml configs as far as possible

Many stock cards still have plenty of options which can only be defined in yaml. My words said a few years ago are still valid. 
Btw, I know only one stock card which got an UI editor for ALL options since I wrote those words. It was a Map card - and it was me who added this editor recently)). I mean - enriching UI editors for stock cards is not trending ))). 

Also, there are three things which I account about UI editor for mini graph card:
1. We need to use some specific controls to input a string, a number; also - to select an entity, a color, a member of some enum. We might use frontend components like “ha-select”, entity selector etc - but there’s a huge risk that next HA version these components will be updated in a way we will not be able to use them. And one of the recent posts in developer blog was about “do not use core components for custom cards”. So, we will have to use OWN components - or stock components with some risks. 
2. When UI is created, there will be an additional task to support changed or new options. 
3. (very selfish) Personally my do not care about UI editor.

ildar170975 committed:
Merge branch 'dev' into editor

ildar170975 committed:
Merge branch 'dev' into editor
